jax/experimental/compilation_cache/gfile_cache.py

```diff
@@ -0,0 +1,48 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pathlib
+
+from jax.experimental.compilation_cache.cache_interface import CacheInterface
+from etils import epath
+from absl import logging
+
+class GFileCache(CacheInterface):
+
+  def __init__(self, path: str):
+    """Sets up a cache at 'path'. Cached values may already be present."""
+    self._path = epath.Path(path)
+    self._path.mkdir(parents=True, exist_ok=True)
+
+  def get(self, key: str):
+    """Returns None if 'key' isn't present."""
+    if not key:
+      raise ValueError("key cannot be empty")
+    path_to_key = self._path / key
+    if path_to_key.exists():
+      return path_to_key.read_bytes()
+    else:
+      return None
+
+  def put(self, key: str, value: bytes):
+    """Adds new cache entry."""
+    if not key:
+      raise ValueError("key cannot be empty")
+    path_to_new_file = self._path / key
+    if 'gs://' in str(path_to_new_file):
+      path_to_new_file.write_bytes(value)
+    else:
+      tmp_path = self._path / f"_temp_{key}"
+      tmp_path.write_bytes(value)
```

Comment:
We should ideally be using the logic from the deleted FileSystemCache to atomically write local files:
```python
          with open(temp_path_to_file, "wb") as file:
              file.write(value)
              file.flush()
              os.fsync(file.fileno())
```

See https://stackoverflow.com/questions/2333872/how-to-make-file-creation-an-atomic-operation. However, this is only valid if we know that self._path refers to a local filesystem -- it could still be `s3://` or something. I think we basically need a `elif <self._path is a local path>`, and then `else` with the current fallback logic, but I can't find a way to determine if self._path is a local path... maybe for now use `'://' not in str(path_to_new_file)`, and we can file an issue against etils to add a way to get the filesystem more robustly?

Reply:
we can have `file:///tmp/x` as the input, which has that `://` in it?  
and are you suggesting I don't use epath for the local file?

Reply:
epath handles local files too :)

Reply:
epath handles local files, but I don't think it does the `flush` + `fsync` that's required for atomicity (there's a performance cost). So I'm saying we use `epath` for everything except writing the file, where we use raw filesystem operations.

Good point about `file://` being valid in addition to not providing a filesystem prefix at all... maybe use `if '://' not in str(path_to_new_file) or 'file://' in str(path_to_new_file)`?

Reply:
We should also make a FR for epath to handle atomic writes :)